### PR TITLE
Add reuse of the latest task definition to `ECSTask`

### PR DESCRIPTION
**Pull request by madkinsz**
This feels really gross, but has been demanded. When a task definition does not match, we will check if the latest task definition in the family meets the requirements. If so, we will use it. This is a little brittle, as AWS sets fields on task definitions on registration, so what you read back can be different. Additionally, if the latest task definition is moved to INACTIVE before the task run is requested, deployment of infrastructure will fail.

Generally, this will require you to set `auto_deregister_task_definition` to `False`.

This pull request is best reviewed [with the whitespace diff hidden](https://github.com/PrefectHQ/prefect-aws/pull/181/files?diff=unified&w=1)

### Checklist

- [ ] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-aws/issues/new/choose) first.
- [ ] Includes tests or only affects documentation.
- [x] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [ ] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [ ] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-aws/blob/main/CHANGELOG.md)

### CODE REVIEW

File: prefect_aws/ecs.py [hunk 1]
```diff
@@ -750,6 +774,58 @@ def _watch_task_and_get_exit_code(
 
         return status_code
 
+    def _task_definitions_equal(self, taskdef_1, taskdef_2) -> bool:
+        """
+        Compare two task definitions.
+
+        Since one may come from the AWS API and have populated defaults, we do our best
+        to homogenize the definitions without changing their meaning.
+        """
```
Comment (madkinsz) [inline prefect_aws/ecs.py:783]: Notably, this part feels like it's going to be brittle and break on edge cases.

File: prefect_aws/ecs.py [hunk 2]
```diff
@@ -750,6 +774,58 @@ def _watch_task_and_get_exit_code(
...
+            """Recursively drop keys with 'empty' values"""
+            for key, value in tuple(dict_.items()):
+                if not value:
+                    dict_.pop(key)
+                if isinstance(value, dict):
+                    _drop_empty_keys(value)
+                if isinstance(value, list):
+                    for v in value:
+                        if isinstance(v, dict):
+                            _drop_empty_keys(v)
```
Comment (madkinsz) [inline prefect_aws/ecs.py:817]: Here we are handling cases like

```python

{
    "containerDefinitions": [
        {
            "name": "prefect",
            "image": "prefecthq/prefect:dev-python3.8",
            "cpu": 1024,
            "memory": 2048,
            "portMappings": [],
            "essential": True,
            "environment": [],
            "mountPoints": [],
            "volumesFrom": [],
        }
    ],
    "family": "test",
    "networkMode": "bridge",
    "volumes": [],
    "placementConstraints": [],
}

# vs

{
    "containerDefinitions": [
        {
            "name": "prefect",
            "image": "prefecthq/prefect:dev-python3.8",
            "cpu": 1024,
            "memory": 2048,
            "essential": True,
        }
    ],
    "family": "test",
    "networkMode": "bridge",
}
```

**Review by desertaxle — APPROVED** [review]
LGTM!

**Comment by madkinsz** [comment]
I'm not sure we should merge this without additional QA. I would not be surprised if there are cases outside those captured by the moto mocks that are relevant.

**Comment by anna-geller** [comment]
sorry for being so slow, QAed just now and it works as expected ✅

image

**Comment by anna-geller** [comment]
Cases I QAed so far:

1. **Rerunning the same block configuration twice** results only in task registration during the first run, 2nd run doesn't rereigster as expected
2. Changing **image** reregisters only during the first run as expected
3. Changing **memory** reregisters only during the first run as expected
4. Changing **cpu** reregisters only during the first run as expected
5. Changing **cluster** reregisters only during the first run as expected
6. Changing **family** reregisters only during the first run as expected
7. Changing **stream_output** reregisters only during the first run as expected
8. Adding a new environment variable reregisters only during the first run as expected
9. Changing `task_start_timeout_seconds` does NOT reregister as expected (Prefect specific argument)
10. Adding task_customizations (`assignPublicIp` DISABLED) reregisters only during the first run as expected
11. Changing `awsvpcConfiguration/subnets` customization does NOT reregister as expected (run-task API specific argument, so doesn't need registration, only changes to which subnet an already registered task gets deployed to)

TL;DR, everything works so far, I think the EC2 launch configuration is the only one missing QA - I won't get to it this week, sorry

The only thing I found surprising is that I somehow got an issue with log stream (nr 7) even though I set the argument configure_cloudwatch_logs to True and set IAM permission to create the log stream.

image

**Comment by madkinsz** [comment]
Re the missing log stream — I think that AWS sometimes lies to us and says it does not exist? I've seen that one in the wild before and have struggled to get a user to open a report. There's an issue at https://github.com/PrefectHQ/prefect-aws/issues/177

Thanks for doing QA!
